### PR TITLE
Authenticated rate limit

### PR DESCRIPTION
**Pull request by eguzki**
Authenticated basic rate limit (after authentication) for all endpoints exposed by the underlying APIs. All endpoints share the same bucket, so if rate limit exceeded for one endpoint, it will also exceed for the remaining endpoints of the same APIProduct.

```yaml
apiVersion: networking.kuadrant.io/v1beta1
kind: APIProduct
metadata:
  name: animaltoys
spec:
  information:
    description: My super nice API Product.
    owner: whoever@mycompany.com
  routing:
    hosts:
      - api.animaltoys.127.0.0.1.nip.io
    expose: true
  APIs:
    - name: dogs
      namespace: default
      tag: production
      mapping:
        prefix: /dogs
    - name: cats
      namespace: default
      tag: production
      mapping:
        prefix: /cats
  securityScheme:
    - name: MyAPIKey
      apiKeyAuth:
        location: authorization_header
        name: APIKEY
        credential_source:
          labelSelectors:
            secret.kuadrant.io/managed-by: authorino
            api: animaltoys
  rateLimit:
    authenticated:
      max_value: 5
      period_in_seconds: 30
    preAuthenticated:
      max_value: 10
      period_in_seconds: 60
```

Implementations details:
* It reconcilliates changes in the `rateLimit.authenticated` object. Removing the `rateLimit.authenticated` section, the controller will disable auth rate limit.
* It is specifically designed to be applied **after** auth http filter.
* Rate limit is based on `user_id`
* Integration with APIKey and OIDC authN
* [Global rate limit](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/global_rate_limiting)
* [Limitador](https://github.com/3scale-labs/limitador) as the gRPC rate limiting service
* Upgraded Authorino to v0.4.0

Depends on #46

### CODE REVIEW

File: apis/networking/v1beta1/apiproduct_types.go [hunk 1]
```diff
@@ -89,11 +89,26 @@ func (a *APISelector) APINamespacedName() types.NamespacedName {
...
 
-type InfraRateLimitSpec struct {
+type AuthRateLimitSpec struct {
 	MaxValue int32 `json:"max_value"`
 	Period   int32 `json:"period_in_seconds"`
 }
 
+type PreAuthRateLimitSpec struct {
+	MaxValue int32 `json:"max_value"`
+	Period   int32 `json:"period_in_seconds"`
```
Comment (guicassolato) [inline apis/networking/v1beta1/apiproduct_types.go:99]: I've left a couple comments about these in #46.

File: apis/networking/v1beta1/apiproduct_types.go [hunk 2]
```diff
@@ -89,11 +89,26 @@ func (a *APISelector) APINamespacedName() types.NamespacedName {
...
+}
+
+type RateLimitSpec struct {
+	// PreAuthRateLimitSpec configures rate limit parameters per remote address
+	// +optional
+	PreAuthRateLimit *PreAuthRateLimitSpec `json:"preAuthenticated,omitempty"`
```
Comment (guicassolato) [inline apis/networking/v1beta1/apiproduct_types.go:105]: I think we should call it something else. "PreAuthRateLimit" makes sense if know that it opposes to "AuthenticatedRateLimit". If the user is planning on using "auth", speaking of "pre-auth" sounds weird IMO.

Maybe we should call this one just "RateLimit"?

I also like the idea (once suggested by @jmprusi) of defining "phases".

File: apis/networking/v1beta1/apiproduct_types.go (hunk 1 above)
Comment (guicassolato) [reply apis/networking/v1beta1/apiproduct_types.go:99]: Can we reuse the struct already defined?

Comment (eguzki) [reply apis/networking/v1beta1/apiproduct_types.go:99]: Sure, I will propagate the changes

**Comment by eguzki** [comment]
Closing in favor of #48

File: apis/networking/v1beta1/apiproduct_types.go (hunk 1 above)
Comment (eguzki) [reply apis/networking/v1beta1/apiproduct_types.go]: > Can we reuse the struct already defined?

we can. I deliberately duplicated the struct to make explicit they are different things and can have different structure and fields. If the design gets stabilized, I will remove duplication (if it exists)

**Pull request closed without merging**